# Homework 5

## Enya Liu

## Logistic Regression without using any libraries. 

In [1]:
# Standard Headers
# You are welcome to add additional headers here if you wish
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Enable inline mode for matplotlib so that Jupyter displays graphs
%matplotlib inline

# Your allowed to use only the above libraries that are imported. No other libs should be used in this assignment. 

## Heart Dataset 

In this Assignment we will work with some patients dataset. 

We have access to 303 patients data. The features are listed below. 

In [2]:
# Your code here
heart_df = pd.read_csv("Heart.csv")
# heart_df = pd.read_csv(r'C:\Users\Ley\Desktop\Heart.csv')
heart_df

Age  Sex     ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  \
0     63    1       typical     145   233    1        2    150      0   
1     67    1  asymptomatic     160   286    0        2    108      1   
2     67    1  asymptomatic     120   229    0        2    129      1   
3     37    1    nonanginal     130   250    0        0    187      0   
4     41    0    nontypical     130   204    0        2    172      0   
..   ...  ...           ...     ...   ...  ...      ...    ...    ...   
298   45    1       typical     110   264    0        0    132      0   
299   68    1  asymptomatic     144   193    1        0    141      0   
300   57    1  asymptomatic     130   131    0        0    115      1   
301   57    0    nontypical     130   236    0        2    174      0   
302   38    1    nonanginal     138   175    0        0    173      0   

     Oldpeak  Slope   Ca        Thal Target  
0        2.3      3  0.0       fixed     No  
1        1.5      2  3.0      normal    Yes  
2        2.6      2  2.0  reversable    Yes  
3        3.5      3  0.0      normal     No  
4        1.4      1  0.0      normal     No  
..       ...    ...  ...         ...    ...  
298      1.2      2  0.0  reversable    Yes  
299      3.4      2  2.0  reversable    Yes  
300      1.2      2  1.0  reversable    Yes  
301      0.0      2  1.0      normal    Yes  
302      0.0      1  NaN      normal     No  

[303 rows x 14 columns]

**Age:** The person’s age in years

**Sex:** The person’s sex (1 = male, 0 = female)

**ChestPain:** chest pain type

* Value 0: asymptomatic
* Value 1: atypical angina
* Value 2: non-anginal pain
* Value 3: typical angina

**RestBP:** The person’s resting blood pressure (mm Hg on admission to the hospital)

**Chol:** The person’s cholesterol measurement in mg/dl

**Fbs:** The person’s fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)
restecg: resting electrocardiographic results

* Value 0: showing probable or definite left ventricular hypertrophy by Estes’ criteria
* Value 1: normal
* Value 2: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)

**RestECG:** The person’s maximum heart rate achieved

**MaxHR:** Exercise induced angina (1 = yes; 0 = no)

**Oldpeak:** ST depression induced by exercise relative to rest (‘ST’ relates to positions on the ECG plot. See more here)

**Slope:** the slope of the peak exercise ST segment — 0: downsloping; 1: flat; 2: upsloping

* 0: downsloping; 
* 1: flat; 
* 2: upsloping

**Ca:** The number of major vessels (0–3)

**Thal:** A blood disorder called thalassemia Value 0: NULL (dropped from the dataset previously

* Value 1: fixed defect (no blood flow in some part of the heart)
* Value 2: normal blood flow
* Value 3: reversible defect (a blood flow is observed but it is not normal)

**Target:** Heart disease (1 = no, 0= yes)

# Task - 1 (4 points)
We want to use logistic regerssion to perdict if the patients will have heart problems or not. The column "Target" in our datasets includes data about heart diseases. If the patient had heart disease we have a 1 and if not a zero. 

Prepare your data set for predicting heart disease ("Target" column) out of 3 features:

* Age of the patient (Column **"Age"**)
* Gender of the patient (male or female - Column **"Sex"**)
* Cholestrol level of the patient (Column **"Chol"**) 


Split your data into 80% traning data and 20% test data, and implement logistic regression model without using any libs than imported above. 

* Do maximum **100 iterations**
* Use a very small learning rate for checking your GD implementation. 
* Your are allowed to use your choice of learning rate, like using 0.0001, 0.001 or 0.01 or 0.1 or higher. 
* Visualize your costs. 
* No need to add an y-intercept in this task. 



In [3]:
# create test/train
train_len = round(len(heart_df)*0.8)

# define X, y
X = np.array(heart_df[['Age','Sex','Chol']])
# convert target from yes no to 1 0
new_target = pd.Series(heart_df['Target'])
y = np.array(new_target.map({'Yes':1,'No':0}))

X_train = X[:train_len]
y_train = y[:train_len]
X_test = X
y_test = y


# logistic regression
def sigmoid(z):
    g = 1 / (1 + np.exp(-z))
    return g

def cost_function(X, y, weights):
    n, d = X.shape
    x_dot_weights = X.dot(weights)
    y = y.astype(np.int)
    cost = 1.0 / n * (-y.T.dot(np.log(sigmoid(x_dot_weights))) - (1 - y).T.dot(np.log(1 - sigmoid(x_dot_weights))))

    return cost

def gradient(X, y, weights):
    n, d = X.shape
    x_dot_weights = X.dot(weights)

    grad = (1.0 / n )* (sigmoid(x_dot_weights) - y).T.dot(X)

    return grad

weights = np.zeros(3)

# Now we optimize it using Gradient Descent. 
num_iterations = 500
learnin_rate = 0.0001

cost_list = []

for i in range(0, num_iterations):
    
    # 
    # Calculate the costs 
    cost = cost_function(X_train, y_train, weights)
    print("Cost is: ", cost)
    # keep the costs for our visualization later 
    
    cost_list.append(cost)
    
    # Calculate the gradients [CODE REMOVED]
    grad = gradient(X_train, y_train, weights)
    
    # Use the gradient to update the weights
    weights = weights - grad*learnin_rate


Cost is:  0.6931471805599453
Cost is:  0.6911184331775377
Cost is:  0.6900362632202808
Cost is:  0.6895345748334202
Cost is:  0.6892692145377417
Cost is:  0.6891346512728148
Cost is:  0.6890589700543245
Cost is:  0.6890140977880899
Cost is:  0.6889837977631849
Cost is:  0.6889609380923364
Cost is:  0.6889417173798884
Cost is:  0.6889243662375394
Cost is:  0.6889079787661342
Cost is:  0.6888921208963703
Cost is:  0.6888765745596712
Cost is:  0.688861233527782
Cost is:  0.6888460447081419
Cost is:  0.6888309817929033
Cost is:  0.6888160314433742
Cost is:  0.6888011867684374
Cost is:  0.6887864440332955
Cost is:  0.6887718010604265
Cost is:  0.6887572564420875
Cost is:  0.6887428091491318
Cost is:  0.6887284583428365
Cost is:  0.6887142032788686
Cost is:  0.6887000432626601
Cost is:  0.6886859776255673
Cost is:  0.6886720057144695
Cost is:  0.6886581268857029
Cost is:  0.6886443405027215
Cost is:  0.6886306459344713
Cost is:  0.6886170425549004
Cost is:  0.6886035297424766
Cost is:  0.688

<ipython-input-3-c74e8ae82c26>:24: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  y = y.astype(np.int)


# Task 2 - (4 points)

Cacluate the accuracy, Precision, Recall and F1 score of your logistic regression implementaion. 
Print the results. 


In [4]:
# Add your code Here! 
def predict(weights, X):
    p = sigmoid(X.dot(weights)) >= 0.5
    return p.astype(int)

predictions = predict(weights, X_test)

# Correct Predictions are the cases that are equal. 

correct = np.sum(predictions == y_test)


print("Accuracy is: ", correct/y_test.size)

# precision = tp/(tp+fp)
# positive = 1
# negative = 0
tp=0
fp=0
fn=0
tn=0
for i in range(0,len(predictions)):
    if predictions[i] == 1 and predictions[i] == y_test[i]:
        tp += 1
    if predictions[i] == 1 and predictions[i] != y_test[i]:
        fp += 1
    if predictions[i] == 0 and predictions[i] != y_test[i]:
        fn += 1
    else:
        tn += 1
precision = tp/(tp+fp)
print('precision is',precision,'for 1')
recall = tp/(tp+fn)
print('recall is',recall,'for 1')
# f-measure
f_measure = 2*precision*recall/(precision+recall)
print('f-measure is',f_measure,'for 1')
print()
precision2 = tn / (tn + fn)
recall2 = tn / (tn + fp)
f_2 = 2 * precision2 * recall2 / (precision2 + recall2)
print('precision is', precision2,'for 0')
print('recall is: ', recall2,'for 0')
print('f-measure is', f_2,'for 0')


Accuracy is:  0.5412541254125413
precision is 0.5 for 1
recall is 0.16546762589928057 for 1
f-measure is 0.24864864864864866 for 1

precision is 0.6171617161716172 for 0
recall is:  0.8904761904761904 for 0
f-measure is 0.7290448343079923 for 0


# Task 3 - (4 points)


Add y-intercept and repeat the above 2 tasks. Do you see any differences after adding the y-intercept? 


In [5]:
# add y-intercept
intercept = np.ones(len(heart_df))
heart_df['yInt']=intercept

# create test/train
train_len = round(len(heart_df)*0.8)

# define X, y
X = np.array(heart_df[['Age','Sex','Chol','yInt']])
# convert target from yes no to 1 0
new_target = pd.Series(heart_df['Target'])
y = np.array(new_target.map({'Yes':1,'No':0}))

X_train = X[:train_len]
y_train = y[:train_len]
X_test = X
y_test = y

weights = np.zeros(4)

# Now we optimize it using Gradient Descent. 
num_iterations = 500
learnin_rate = 0.0001

cost_list = []

for i in range(0, num_iterations):
    
    # 
    # Calculate the costs 
    cost = cost_function(X_train, y_train, weights)
    print("Cost is: ", cost)
    # keep the costs for our visualization later 
    
    cost_list.append(cost)
    
    # Calculate the gradients [CODE REMOVED]
    grad = gradient(X_train, y_train, weights)
    
    # Use the gradient to update the weights
    weights = weights - grad*learnin_rate

<ipython-input-3-c74e8ae82c26>:24: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  y = y.astype(np.int)


Cost is:  0.6931471805599453
Cost is:  0.6911185326894426
Cost is:  0.6900363280289841
Cost is:  0.689534584014115
Cost is:  0.6892691675169696
Cost is:  0.6891345600038361
Cost is:  0.6890588425071298
Cost is:  0.689013940697307
Cost is:  0.688983615036766
Cost is:  0.6889607322474902
Cost is:  0.6889414897981837
Cost is:  0.6889241176997609
Cost is:  0.688907709665768
Cost is:  0.688891831418801
Cost is:  0.6888762647688124
Cost is:  0.6888609034225133
Cost is:  0.6888456942513534
Cost is:  0.6888306109284995
Cost is:  0.6888156401051904
Cost is:  0.6888007748851921
Cost is:  0.6887860115311927
Cost is:  0.6887713478645602
Cost is:  0.6887567824771754
Cost is:  0.6887423143399052
Cost is:  0.6887279426142406
Cost is:  0.6887136665561677
Cost is:  0.6886994854714905
Cost is:  0.6886853986919635
Cost is:  0.6886714055648773
Cost is:  0.6886575054469843
Cost is:  0.6886436977021564
Cost is:  0.6886299816997572
Cost is:  0.6886163568141511
Cost is:  0.688602822424221
Cost is:  0.68858937

In [12]:
predictions = predict(weights, X_test)

# Correct Predictions are the cases that are equal. 

correct = np.sum(predictions == y_test)


print("Accuracy is: ", correct/y_test.size)

# precision = tp/(tp+fp)
# positive = 1
# negative = 0
tp=0
fp=0
fn=0
tn=0
for i in range(0,len(predictions)):
    if predictions[i] == 1 and predictions[i] == y_test[i]:
        tp += 1
    if predictions[i] == 1 and predictions[i] != y_test[i]:
        fp += 1
    if predictions[i] == 0 and predictions[i] != y_test[i]:
        fn += 1
    else:
        tn += 1
precision = tp/(tp+fp)
print('precision is',precision,'for 1')
recall = tp/(tp+fn)
print('recall is',recall,'for 1')
# f-measure
f_measure = 2*precision*recall/(precision+recall)
print('f-measure is',f_measure,'for 1')
print()
precision2 = tn / (tn + fn)
recall2 = tn / (tn + fp)
f_2 = 2 * precision2 * recall2 / (precision2 + recall2)
print('precision is', precision2,'for 0')
print('recall is: ', recall2,'for 0')
print('f-measure is', f_2,'for 0')
print()
print('There is a difference.')
print('After adding the y-intercept, the accuracy, precision and f-feature slightly increases for 1.')

Accuracy is:  0.5412541254125413
precision is 0.5 for 1
recall is 0.02877697841726619 for 1
f-measure is 0.05442176870748299 for 1

precision is 0.5544554455445545 for 0
recall is:  0.9767441860465116 for 0
f-measure is 0.7073684210526316 for 0

There is a difference.
After adding the y-intercept, the accuracy, precision and f-feature slightly increases for 1.


# Task 4 - Implement the Bold Driver   - (4 points)
In your GD implementation, add the bold driver idea to have a dynamic learning rate.

* Add a stop codition to stop the GD when the cost is not changing more than 0.001. (differences between two costs not more than 0.001, then stop)
* Can you stop earlier than 100 iterations? 




In [16]:
# Now we optimize it using Gradient Descent. 
num_iterations = 500
learnin_rate = 0.0001

cost_list = []

# Implementation here is removed. 

# Your task to implement the GD here. 

for i in range(0, num_iterations):
    
    # 
    # Calculate the costs 
    cost = cost_function(X_train, y_train, weights)
    print("Cost is: ", cost)
    # keep the costs for our visualization later 
    
    cost_list.append(cost)
    
    # Calculate the gradients [CODE REMOVED]
    grad = gradient(X_train, y_train, weights)
    weights = weights - grad*learnin_rate
    
    precision = 0.001
    # bold driver
    if i > 0 and np.abs(cost_list[i-1]-cost_list[i]) < precision:
        print('Iteration stopped at',i,'earlier than 100 iterations')
        break
    if i > 0 and cost_list[i-1] > cost_list[i]:
        learnin_rate = 1.05*learnin_rate
    if i > 0 and cost_list[i-1] < cost_list[i]:
        learnin_rate = 0.5 * learnin_rate

    
    old_cost = cost


Cost is:  0.6874507744351089
Cost is:  0.6874448933366729
Iteration stopped at 1 earlier than 100 iterations


<ipython-input-3-c74e8ae82c26>:24: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  y = y.astype(np.int)


# Task 5 - Implement the L2 norm regularization.  - (4 points)

Modify your Cost and gradient to implement the l2 norm regularization. 
Repreat the task 1 and 2 to check if your result is changing. 

* Use y-itercept. 
* Do max 100 iterations as before and report your accuracy, Precision, Recall and F1. 
* You can stop earlier when the cost is not changing than 0.001. 


**Optional:** you might want to use the bold driver. But you can do this task without the bold driver as well. 



In [8]:
# Add your code Here! 
num_iterations = 100
learnin_rate = 0.0001

weights = np.zeros(4)

cost_list = []

lamb = 1
def sigmoid(v):
    return 1/(1+np.e**-v)

def logistic_regression(X, weights):
    pred = np.sum(weights*X)
    return sigmoid(pred)

def mse(y, pred):
    return (-1/len(heart_df)) * sum((y-pred)**2)


for i in range(num_iterations):
    # predictions
    # pred = logistic_regression(X_train, weights)
    
    # calculate MSE
    # cost = mse(y_train, pred)*len(y_train)
    
    # calculate gradient
    cost = cost_function(X_train, y_train, weights)*len(y_train)
    
    # calculate ridge regression/penalty
    ridge_cost = lamb*sum([w**2 for w in weights[:-1]])
    cost += ridge_cost
    print('Cost is', cost)
    
    cost_list.append(cost)
    
    grad = gradient(X_train, y_train, weights)
    ridge_gradient = lamb * sum(weights[:-1])
    weights = weights - learnin_rate * (grad + ridge_gradient)
    
    precision = 0.001
    # bold driver
    if i > 0 and np.abs(cost_list[i-1]-cost_list[i]) < precision:
        print('Iteration stopped at',i)
        break
    if i > 0 and cost_list[i-1] > cost_list[i]:
        learnin_rate = 1.05*learnin_rate
    if i > 0 and cost_list[i-1] < cost_list[i]:
        learnin_rate = 0.5 * learnin_rate

    
    old_cost = cost

Cost is 167.74161769550676
Cost is 167.25068630076058
Cost is 166.9890055644544
Cost is 166.89713551585277
Cost is 166.85716117032223
Cost is 166.84669533898503
Cost is 166.85582969822406
Cost is 166.89634329265928
Cost is 166.730032672741
Cost is 166.72794686667228
Cost is 166.72647725626302
Cost is 166.72509649336138
Cost is 166.72368136824687
Cost is 166.72220460499284
Cost is 166.72065892004343
Cost is 166.71904057631292
Cost is 166.71734629563855
Cost is 166.7155727616543
Cost is 166.713716550295
Cost is 166.7117741201366
Cost is 166.70974181029757
Cost is 166.70761583962397
Cost is 166.70539230646577
Cost is 166.70306718898595
Cost is 166.70063634604023
Cost is 166.69809551869
Cost is 166.69544033242389
Cost is 166.69266630016799
Cost is 166.68976882617122
Cost is 166.68674321085933
Cost is 166.6835846567562
Cost is 166.68028827557708
Cost is 166.6768490966064
Cost is 166.6732620764766
Cost is 166.6695221104712
Cost is 166.66562404548105
Cost is 166.66156269474607
Cost is 166.657

<ipython-input-3-c74e8ae82c26>:24: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  y = y.astype(np.int)


In [17]:
predictions = predict(weights, X_test)

# Correct Predictions are the cases that are equal. 

correct = np.sum(predictions == y_test)


print("Accuracy is: ", correct/y_test.size)

# precision = tp/(tp+fp)
# positive = 1
# negative = 0
tp=0
fp=0
fn=0
tn=0
for i in range(0,len(predictions)):
    if predictions[i] == 1 and predictions[i] == y_test[i]:
        tp += 1
    if predictions[i] == 1 and predictions[i] != y_test[i]:
        fp += 1
    if predictions[i] == 0 and predictions[i] != y_test[i]:
        fn += 1
    else:
        tn += 1
precision = tp/(tp+fp)
print('precision is',precision,'for 1')
recall = tp/(tp+fn)
print('recall is',recall,'for 1')
# f-measure
f_measure = 2*precision*recall/(precision+recall)
print('f-measure is',f_measure,'for 1')
print()
precision2 = tn / (tn + fn)
recall2 = tn / (tn + fp)
f_2 = 2 * precision2 * recall2 / (precision2 + recall2)
print('precision is', precision2,'for 0')
print('recall is: ', recall2,'for 0')
print('f-measure is', f_2,'for 0')
print()
print('After implemented L2 norm regularization, I did not see any change with the accuracy, precision, recall or f-measure.')
print('They are the same as those of logistic regression with y-intercept and bold driver. ')

Accuracy is:  0.5412541254125413
precision is 0.5 for 1
recall is 0.02877697841726619 for 1
f-measure is 0.05442176870748299 for 1

precision is 0.5544554455445545 for 0
recall is:  0.9767441860465116 for 0
f-measure is 0.7073684210526316 for 0

After implemented L2 norm regularization, I did not see any change with the accuracy, precision, recall or f-measure.
They are the same as those of logistic regression with y-intercept and bold driver. 
